README.md
# time-series-library

A TypeScript/JavaScript library for arithmetic on Statistics Canada time series 
vector data.

This library can be used with Node.js or any of the following web browsers:
- Chrome
- Firefox
- Internet Explorer 11
- Opera
- Safari

# Installation

Using Node.js:
```javascript
// common-js
const TimeSeriesLibrary = require('time-series-library');
const Vector = TimeSeriesLibrary.Vector;
const VectorLib = TimeSeriesLibrary.VectorLib;

// es6 module
import {Vector, VectorLib} from 'time-series-library';
```

Using a web browser:
```html
<script src="./time-series-library.js"></script>

<script>
    const Vector = TimeSeriesLibrary.Vector;
    const VectorLib = TimeSeriesLibrary.VectorLib;
</script>
```

# Usage

**Vector:**  
[get(index)](#Vector.get)  
[refper(index), refperStr(index)](#Vector.refper)  
[value(index)](#Vector.value)  
[values()](#Vector.values)  
[length](#Vector.length)  
[push(datapoint)](#Vector.push)  
[equals(other)](#Vector.equals)  
[copy()](#Vector.copy)  
[map(fn)](#Vector.map)  
[find(predicate)](#Vector.find)  
[filter(predicate)](#Vector.filter)  
[range(startDate, endDate)](#Vector.range)  
[latestN(n)](#Vector.latestN)  
[interoperable(other)](#Vector.interoperable)  
[intersection(others)](#Vector.intersection)  
[sum()](#Vector.sum)  
[average()](#Vector.average)  
[reduce()](#Vector.reduce)  
[operate(operation)](#Vector.operate)  
[periodDeltaTransformation(transformation)](#Vector.periodDeltaTransformation)  
[samePeriodPreviousYearTransformation(transformation)](#Vector.samePeriodPreviousYearTransformation)  
[periodTransformation(transformation)](#Vector.periodTransformation)  
[periodToPeriodPercentageChange(transformation)](#Vector.periodToPeriodPercentageChange)  
[periodToPeriodDifference()](#Vector.periodToPeriodDifference)  
[samePeriodPreviousYearPercentageChange()](#Vector.samePeriodPreviousYearPercentageChange)  
[samePeriodPreviousYearDifference()](#Vector.samePeriodPreviousYearDifference)  
[convertToFrequency(mode, converter)](#Vector.convertToFrequency)  
[weekly(mode)](#Vector.weekly)  
[monthly(mode)](#Vector.monthly)  
[biMonthly(mode)](#Vector.biMonthly)  
[quarterly(mode)](#Vector.quarterly)  
[semiAnnual(mode)](#Vector.semiAnnual)  
[annual(mode)](#Vector.annual)  
[biAnnual(mode)](#Vector.biAnnual)  
[triAnnnual(mode)](#Vector.triAnnual)  
[quadrennial(mode)](#Vector.quadrennial)  
[quinquennial(mode)](#Vector.quinquennial)  
[round(decimals)](#Vector.round)  
[roundBankers(decimals)](#Vector.roundBankers)  
[json()](#Vector.json)  

**VectorLib:**  
[evaluate(expression, vectors)](#VectorLib.evaluate)  
[generateDaily(values, startDate)](#VectorLib.generateDaily)  
[generateWeekly(values, startDate)](#VectorLib.generateWeekly)  
[generateMonthly(values, startDate)](#VectorLib.generateMonthly)  
[generateBiMonthly(values, startDate)](#VectorLib.generateBiMonthly)  
[generateQuarterly(values, startDate)](#VectorLib.generateQuarterly)  
[generateSemiAnnual(values, startDate)](#VectorLib.generateSemiAnnual)  
[generateAnnual(values, startDate)](#VectorLib.generateAnnual)  
[generateBiAnnual(values, startDate)](#VectorLib.generateBiAnnual)  
[generateTriAnnual(values, startDate)](#VectorLib.generateTriAnnual)  
[generateQuadrennial(values, startDate)](#VectorLib.GenerateQuadrennial)  
[generateQuinquennial(values, startDate)](#VectorLib.generateQuinquennial)  

## Vector

To create a new vector:
```javascript
let vector = new Vector();
```

Vectors can also be initialized with data:
```javascript
let vector = new Vector([
    {'refper': "2018-01-01", 'value': 1},
    {'refper': "2018-02-01", 'value': 2},
    {'refper': "2018-03-01", 'value': 3}
]);
```

The value of **refper** may be a **yyyy-mm-dd** formatted date string or a 
**Date** object.

Datapoints may also hold addional information in the `metadata property`:
```js
let vector = let vector = new Vector([
    {'refper': "2018-01-01", 'value': 1, 'metadata': {'key': 'value'}}
]); 
```

<a name="Vector.get"></a>
### get(index)

Returns the datapoint of a vector at a specific index.

Example:
```javascript
let vector = new Vector([
    {'refper': "2018-01-01", 'value': 1},
    {'refper': "2018-02-01", 'value': 2},
    {'refper': "2018-03-01", 'value': 3}
]);

let result = vector.get(1);
```

Result:
```javascript
{'refper': "2018-02-01", 'value': 2}
```

<a name="Vector.refper"></a>
### refper(index), refperStr(index)

Gets the refeperence period of a datapoint at a specific index. 

The function **refper** returns a data object and the function **refperStr** 
returns a **yyyy-mm-dd** formatted date string. 

Example:
```javascript
let vector = new Vector([
    {'refper': "2018-01-01", 'value': 1},
    {'refper': "2018-02-01", 'value': 2},
    {'refper': "2018-03-01", 'value': 3}
]);

let result = vector.refperStr(1);
```

Result:
```javascript
"2018-02-01"
```

<a name="Vector.value"></a>
### value(index)

Return the value of a datapoint at a specific index.

Example:
```javascript
let vector = new Vector([
    {'refper': "2018-01-01", 'value': 1},
    {'refper': "2018-02-01", 'value': 2},
    {'refper': "2018-03-01", 'value': 3}
]);

let result = vector.value(1);
```

Result:
```javascript
2
```

<a name="Vector.values"></a>
### values()

Return the list of values in a vector.

Example:
```javascript
let vector = new Vector([
    {'refper': "2018-01-01", 'value': 1},
    {'refper': "2018-02-01", 'value': 2},
    {'refper': "2018-03-01", 'value': 3}
]);

let result = vector.values();
```

Result:
```javascript
[1, 2, 3]
```

<a name="Vector.length"></a>
### length

Gets the length of a vector.

Example:
```javascript
let vector = new Vector([
    {'refper': "2018-01-01", 'value': 1},
    {'refper': "2018-02-01", 'value': 2}
]);

let result = vector.length;
```

Result:
```javascript
2
```

<a name="Vector.push"></a>
### push(datapoint)

Appends a datapoint to a vector.

Example:
```javascript
let vector = new Vector([
    {'refper': "2018-01-01", 'value': 1}
]);

vector.push({'refper': "2018-02-01", 'value': 2});
```

Result:
```javascript
[
    {'refper': "2018-01-01", 'value': 1},
    {'refper': "2018-02-01", 'value': 2}
]
```

<a name="Vector.equals"></a>
### equals(other)

Checks if two vectors are equal.

Two vectors **a** and **b** are equal if the number of datapoints in **a** is 
equal to the number of datapoints in **b**, and for each datapoint with 
reference period **Ra** and value **Va** in **a**, there exists a datapoint in 
**b** whose reference period is equal to **Ra** and value is equal to **Va**. 

Example:
```javascript
let v1 = new Vector([
    {'refper': "2018-01-01", 'value': 1},
    {'refper': "2018-02-01", 'value': 2}
]);
let v2 = new Vector([
    {'refper': "2018-01-01", 'value': 1},
    {'refper': "2018-02-01", 'value': 2}
]);

let result = v1.equals(v2);
```

Result:
```javascript
true
```

<a name="Vector.copy"></a>
### copy()

Returns a new **Vector** object that is a copy of the calling **Vector**.

Example:
```javascript
let vector = new Vector([
    {'refper': "2018-01-01", 'value': 1},
    {'refper': "2018-02-01", 'value': 2}
]);

let result = vector.copy();
```

Result:
```javascript
[
    {'refper': "2018-01-01", 'value': 1},
    {'refper': "2018-02-01", 'value': 2}
]
```

<a name="Vector.map"></a>
### map(fn)

Applies a map function, **fn** on each point in a vector.   

Example:
```javascript
let vector = new Vector([
    {'refper': "2018-01-01", 'value': 0},
    {'refper': "2018-01-02", 'value': 1},
    {'refper': "2018-01-03", 'value': 2},
]);

let result = vector.map(p => p.value);
```

Result:
```javascript
[0, 1, 2]
```

<a name="Vector.find"></a>
### find(predicate)

Returns the first datapoint in a vector matching a predicate function. Returns 
`undefined` if no match is found.   

Example:
```javascript
let vector = new Vector([
    {'refper': "2018-01-01", 'value': 0},
    {'refper': "2018-01-02", 'value': 1},
    {'refper': "2018-01-03", 'value': 2},
]);

let result = vector.find(p => p == 1);
```

Result:
```javascript
{'refper': "2018-01-02", 'value': 1}
```

<a name="Vector.filter"></a>
### filter(predicate)

Returns a filtered vector based on a predicate function. The function 
**predicate** should be a function that operates on a datapoint and returns 
a **boolean**.   

Example:
```javascript
let vector = new Vector([
    {'refper': "2018-01-01", 'value': 0},
    {'refper': "2018-01-02", 'value': 1},
    {'refper': "2018-01-03", 'value': 2},
]);

let result = vector.filter(p => p.value % 2 == 0);
```

Result:
```javascript
[
    {'refper': "2018-01-01", 'value': 0},
    {'refper': "2018-01-03", 'value': 2}
]
```

<a name="Vector.range"></a>
### range(startDate, endDate)

Returns the vector constrained within a specified range. The parameters 
**startDate** and **endDate** may be either date strings or date objects.

Example:
```javascript
let vector = new Vector([
    {'refper': "2018-01-01", 'value': 0},
    {'refper': "2018-02-01", 'value': 1},
    {'refper': "2018-03-01", 'value': 2},
    {'refper': "2018-04-01", 'value': 3}
]);

let result = vlib.range("2018-02-01", "2018-03-01");
```

Result:
```javascript
[
    {'refper': "2018-02-01", 'value': 1},
    {'refper': "2018-03-01", 'value': 2}
]
```

<a name="Vector.range"></a>
### latestN

Returns a new vector containing only the last n reference periods of the 
calling vector.

Example:
```javascript
let vector = new Vector([
    {'refper': "2018-01-01", 'value': 0},
    {'refper': "2018-02-01", 'value': 1},
    {'refper': "2018-03-01", 'value': 2},
    {'refper': "2018-04-01", 'value': 3}
]);

let result = vlib.latestN(2);
```

Result:
```javascript
[
    {'refper': "2018-03-01", 'value': 2},
    {'refper': "2018-04-01", 'value': 3}
]
```

<a name="Vector.interoperable"></a>
### interoperable(other)

Checks if two vectors are interoperable.

Two vectors **a** and **b** are interoperable if the number of datapoints in 
**a** is equal to the number of datapoints in **b**, and for each datapoint with 
reference period **Ra** in **a**, there exists a datapoint in **b** whose 
reference period is equal to **Ra**.

Example:
```javascript
let v1 = new Vector([
    {'refper': "2018-01-01", 'value': 1},
    {'refper': "2018-02-01", 'value': 2}
]);
let v2 = new Vector([
    {'refper': "2018-01-01", 'value': 3},
    {'refper': "2018-02-01", 'value': 4}
]);

let result = v1.interoperable(v2);
```

Result:
```javascript
true
```

<a name="Vector.intersection"></a>
### intersection(others)

Returns the result of the intersection of the calling vector with other vectors. 

The parameter **others** can either be a single vector or an array of vectors.

The result of vectors **a** intersected with **b** is defined as the vector 
**a'** and such that **a'** is composed only of datapoints with reference 
periods defined in both **a** and **b**.

Example:
```javascript
let v1 = [
    {'refper': "2018-01-01", 'value': 1},
    {'refper': "2018-02-01", 'value': 2},
    {'refper': "2018-03-01", 'value': 3},
    {'refper': "2018-04-01", 'value': 4}
];
let v2 = [
    {'refper': "2018-01-01", 'value': 5},
    {'refper': "2018-03-01", 'value': 6}
];

let result = v1.intersection(v2);
```

Result:
```javascript
[
    {'refper': "2018-01-01", 'value': 1},
    {'refper': "2018-03-01", 'value': 3}
]
```

<a name="Vector.sum"></a>
### sum()

Returns the sum of all values in a vector.

Example:
```javascript
let vector = new Vector([
    {'refper': "2018-01-01", value: 1},
    {'refper': "2018-02-01", value: 2},
    {'refper': "2018-03-01", value: 3},
]);

let result = vector.sum();
```

Result:
```javascript
6
```

<a name=Vector.average></a>
### average()

Returns the average of the values in a vector.

Example:
```javascript
let vector = new Vector([
    {'refper': "2018-01-01", value: 1},
    {'refper': "2018-02-01", value: 2},
    {'refper': "2018-03-01", value: 3},
]);

let result = vector.average();
```

Result:
```javascript
2
```

<a name="Vector.reduce"></a>
### reduce(reducer)

Reduces the values in a vector based on the function **reducer**.

The function **reducer** has Number parameters **accumulator** and **current** 
and should return a Number.

Example:
```javascript
let vector = new Vector([
    {'refper': "2018-01-01", 'value': 2},
    {'refper': "2018-02-01", 'value': 3},
    {'refper': "2018-02-01", 'value': 4}
]);  
let result = vector.reduce(function(accumulator, current) {
    return accumulator * current;
});
```

Result:
```
24
```

<a name="Vector.operate"></a>
### operate(operation)

Performs an operation with another vector. The parameter **operation** is 
a function with two **Number** inputs **a** and **b** that returns the result 
of an operation on both inputs. The input **a** should represent the value in 
the calling vector and the input **b** should represent the value in the 
other vector.

The intersection of the two vectors are taken before performing the operation.

Example:
```js
let vectorA = new Vector([
    {'refper': "2018-01-01", 'value': 1},
    {'refper': "2018-02-01", 'value': 2}
]);

let vectorB = new Vector([
    {'refper': "2018-01-01", 'value': 3},
    {'refper': "2018-02-01", 'value': 4}
]);

let result = vectorA.operate(vectorB, (a, b) => a + b);
```

Result:
```js
[
    {'refper': "2018-01-01", 'value': 4},
    {'refper': "2018-02-01", 'value': 6}  
]
```

<a name="Vector.periodDeltaTransformation"></a>
### periodDeltaTransformation(transformation)

Returns a new vector that is the result of a period to period transformation on 
the calling vector. The parameter **transformation** should be a function 
with two **Number** inputs representing the value of the current period and 
previous period and should return the transformed value.

The first datapoint of the transformed vector will have a value of `null` since 
there is no previous period to compare to.

Example:
```javascript
let vector = new Vector([
    {'refper': "2018-01-01", value: 1},
    {'refper': "2018-02-01", value: 2},
    {'refper': "2018-03-01", value: 3},
]);

let result = vector.periodDeltaTransformation(function(curr, last)  {
    return curr + last;
});  
```

Result:
```javascript
[
    {'refper': "2018-01-01", value: null},
    {'refper': "2018-02-01", value: 3},
    {'refper': "2018-03-01", value: 5}
]
```

<a name="Vector.samePeriodPreviousYearTransformation"></a>
### samePeriodPreviousYearTransformation(transformation)

Returns a new vector that is the result of a same period previous year 
transformation on the calling vector. The parameter **transformation** should 
be a function with two **Number** inputs representing the value of the current 
period and previous period and should return the transformed value.

The first few datapoints of the transformed vector will have a value of `null` 
since there is no previous year to compare to.

Example:
```javascript
let vector = new Vector([
    {'refper': '2018-03-31', 'value': 1},
    {'refper': '2018-06-30', 'value': 2},
    {'refper': '2018-09-30', 'value': 3},
    {'refper': '2018-12-31', 'value': 4},
    {'refper': '2019-03-31', 'value': 5},
    {'refper': '2019-06-30', 'value': 6},
    {'refper': '2019-09-30', 'value': 7},
    {'refper': '2019-12-31', 'value': 8}
]);

let result = vector.samePeriodPreviousYearTransformation(function(curr, last)  {
    return curr - last;
});  
```

Result:
```javascript
[
    {'refper': '2018-03-31', 'value': null},
    {'refper': '2018-06-30', 'value': null},
    {'refper': '2018-09-30', 'value': null},
    {'refper': '2018-12-31', 'value': null},
    {'refper': '2019-03-31', 'value': 4},
    {'refper': '2019-06-30', 'value': 4},
    {'refper': '2019-09-30', 'value': 4},
    {'refper': '2019-12-31', 'value': 4}
]
```

<a name="Vector.periodTransformation"></a>
### periodTransformation(transformation)

Returns a new vector that is the result of period transformation on 
the calling vector. The parameter **transformation** should be a function 
with a **Number** input representing the value of the current period and should 
return the transformed value.

The first datapoint of the transformed vector will have a value of `null` since 
there is no previous period to compare to.

Example:
```javascript
let vector = new Vector([
    {'refper': "2018-01-01", value: 1},
    {'refper': "2018-02-01", value: 2},
    {'refper': "2018-03-01", value: 3}
]);

let result = vector.periodTransformation(function(value)  {
    return value * 2;
});  
```

Result:
```javascript
[
    {'refper': "2018-01-01", value: 2},
    {'refper': "2018-02-01", value: 4},
    {'refper': "2018-03-01", value: 6}
]
```

<a name="Vector.periodToPeriodPercentageChange"></a>
### periodToPeriodPercentageChange()

Returns a period-to-period percentage change vector of the calling vector.

Example:
```javascript
let vector = new Vector([
    {'refper': '2018-01-01', 'value': 2},
    {'refper': '2018-02-01', 'value': 6},
    {'refper': '2018-03-01', 'value': 3}
]);

let result = vector.periodToPeriodPercentageChange();
```

Result:
```javascript
[
    {'refper': "2018-01-01", 'value': null},
    {'refper': "2018-02-01", 'value': 200.0},
    {'refper': "2018-03-01", 'value': -50.0}
]
```

<a name="Vector.periodToPeriodDifference"></a>
### periodToPeriodDifference()

Returns a period-to-period difference vector of the calling vector.

Example:
```javascript
let vector = new Vector([
    {'refper': '2018-01-01', 'value': 2},
    {'refper': '2018-02-01', 'value': 6},
    {'refper': '2018-03-01', 'value': 3}
]);

let result = vector.periodToPeriodDifference();
```

Result:
```javascript
[
    {'refper': "2018-01-01", 'value': null},
    {'refper': "2018-02-01", 'value': 4},
    {'refper': "2018-03-01", 'value': -3}
]
```

<a name="Vector.samePeriodPreviousYearPercentageChange"></a>
### samePeriodPreviousYearPercentageChange()

Returns the same period previous year percentage change vector of the calling 
vector.

**Note:** Only works for vectors with frequencies of monthly of lower.

Example:
```javascript
let vector = new Vector([
    {'refper': '2018-06-01', 'value': 1},
    {'refper': '2018-12-01', 'value': 2},
    {'refper': '2019-06-01', 'value': 4},
    {'refper': '2019-12-01', 'value': 8},
    {'refper': '2020-06-01', 'value': 4},
    {'refper': '2020-12-01', 'value': 4}
]);

let result = vector.samePeriodPreviousYearPercentageChange();
```

Result:
```javascript
[
    {'refper': '2018-06-01', 'value': null},
    {'refper': '2018-12-01', 'value': null},
    {'refper': '2019-06-01', 'value': 300.0},
    {'refper': '2019-12-01', 'value': 300.0},
    {'refper': '2020-06-01', 'value': 0},
    {'refper': '2020-12-01', 'value': -50.0}
]
```

<a name="Vector.samePeriodPreviousYearDifference"></a>
### samePeriodPreviousYearDifference()

Returns the same period previous year difference vector of the calling 
vector.

**Note:** Only works for vectors with frequencies of monthly of lower.

Example:
```javascript
let vector = new Vector([
    {'refper': '2018-06-01', 'value': 0},
    {'refper': '2018-12-01', 'value': 2},
    {'refper': '2019-06-01', 'value': 0},
    {'refper': '2019-12-01', 'value': 6},
    {'refper': '2020-06-01', 'value': 0},
    {'refper': '2020-12-01', 'value': 4}
]);

let result = vector.samePeriodPreviousYearPercentageChange();
```

Result:
```javascript
[
    {'refper': '2018-06-01', 'value': null},
    {'refper': '2018-12-01', 'value': null},
    {'refper': '2019-06-01', 'value': 0},
    {'refper': '2019-12-01', 'value': 4},
    {'refper': '2020-06-01', 'value': 0},
    {'refper': '2020-12-01', 'value': -2}
]
```

<a name="Vector.convertToFrequency"></a>
### convertToFrequency(mode, converter)

Converts the frequency of a vector to a user defined frequency.

The parameter **mode** is optional and can be one of the following strings:
- `"last"`: Takes the last reference period of each user defined frequency 
(Default).
- `"sum"`: Takes the sum of each user defined frequency.
- `"average"`: Takes the average of each user defined frequency.
- `"max"`: Takes the maximum value of each user defined frequency.
- `"min"`: Takes the minimum value of each user defined frequency.

The parameter **conveter** is a function with date parameters **curr** and 
**last** and a `Boolean` return value. This function should return true when 

CHANGELOG.md
# Changelog

## [3.2.0] - 2019-09-13
### Added
- Created namespace containing vector Point interfaces.

## [3.1.0] - 2019-07-29
### Added
- Added quadrennial() and generateQuadrennial() functions.

## [3.0.2] - 2019-07-18
### Fixed
- Fixed reading of vector expression symbols for symbols longer than one 
character.

## [3.0.1] - 2019-07-17
### Added
- Added TypeScript docstrings.

## [3.0.0] - 2019-07-17
### Changed
- Library rewrite in TypeScript.
- Extra data in points is now attached to the `metadata` property.

## [2.6.2] - 2019-06-27
### Fixed
- Fixed characters before first vector identifier in vector expression being 
parsed as a vector ID.

## [2.6.1] - 2019-06-21
### Fixed
- Fixed same period previous year transformations not correctly recognizing the 
previous February reference period.

## [2.6.0] - 2019-06-19
### Added
- Added biMonthly() and triAnnual() functions.
- Added generateBiMonthly() and generateTriAnnual() functions.

### Chanegd
- Allow other arguments to map, reduce, and filter functions.

### Fixed
- Fixed handling of dates to be compatible with all time zones. 

## [2.5.4] - 2019-06-10
### Changed
- Removed 'v' character from vector identifiers.

## [2.5.3] - 2019-06-04
### Changed
- Fixed computation of last frequency chunk for annual conversions.

## [2.5.2] - 2019-05-29
### Changed
- Frequecy conversions to annual are now always fixed to the last reference 
period of each year.

## [2.5.1] - 2019-05-28
### Changed
- Replaced references to tetra-annual to quinquennial.

## [2.5.0] - 2019-05-28
### Added
- Added values() function.
- Added user defined convertToFrequency() function.
- Added semiAnnual() frequency conversion function.
- Added biAnnual() frequency conversion function.
- Added tetraAnnual() frequency conversion function.
- Added min and max modes for frequency conversion functions.

### Changed
- Frequnecy conversion functions now start from the lastest reference period 
and progress in descending order of time.

## [2.4.0] - 2019-05-24
### Added
- Added vector generation functions.
- Same period previous year transformation function.

### Changed
- Corrected implementation of same period previous year functions.

## [2.3.0] - 2019-04-11
### Added
- Added weekly() function.
- Added quarterly() alias to quarter().
- Added json() function.
- Added map() function.
- Added find() function.

## [2.2.0] - 2019-02-20
### Added
- Added quarter() function.

## [2.1.0] - 2019-02-07
### Added
- Added sum() function.
- Added average() function.
- Added reduce() function.
- Added operate() function.
- Added annual() function.
- Added monthly() function.

## [2.0.0] - 2019-01-22
### Added
- Added Vector class. Most functions are now functions of this class.
- latestN() function.
- get() function.
- value() function.
- refper() function.
- Support for Internet Explorer 11.

### Changed
- VectorLib was seperated into Vector and VectorLib.
- round() and roundBankers() now return a new vector and no longer operate in 
place.
- Vectors will now always use date objects internally instead of date strings.
- Vectors use local dates for reference periods instead of UTC.

### Removed
- Bulk intersection function from VectorLib. This has been replaced by the 
intersection function of Vector.

## [1.3.0] - 2018-12-27
### Added
- periodTransformation() function.
- samePeriodPreviousYearPercentageChange() function.
- samePeriodPreviousYearDifference function().
- annualize() function.

### Changed 
- Intersection of vectors is now applied before any vector operation.

## [1.2.0] - 2018-12-11
### Added
- range() function.

## [1.1.0] - 2018-11-27
### Added
- Period-to-period difference and percentage change functions.
- filter() function.
- Default and Banker's rounding algorithms.

vexp_parser.js
VectorExpressionEvaluator = function() {
    operators = {
        '+': function(a, b) { return a + b; },
        '-': function(a, b) { return a - b; },
        '*': function(a, b) { return a * b; },
        '/': function(a, b) { return a / b; }
    };
    
    operatorPriorities = {
        '*': 2,
        '/': 2,
        '+': 1,
        '-': 1,
    };
    
    
    this.equals = function(vectorA, vectorB) {
        if (vectorA.length != vectorB.length) {
            return false;
        }
        
        for (let p = 0; p < vectorA.length; p++) {
            if (vectorA[p].refper != vectorB[p].refper) {
                return false;
            }
            if (vectorA[p].value != vectorB[p].value) {
                return false;
            }
        }
        
        return true;
    }
    
    
    this.getVectorIds = function(expression) {
        expression = expression.replace(/ /g, '');
        let ids = [];	
        let nextId = "";
        for (let c = 0; c < expression.length; c++) {
            if (!isNaN(expression[c])) {
                nextId += expression[c];
            } else {
                if (nextId != "") ids.push(nextId);
                nextId = "";
            }
        }
        
        if (nextId != "") ids.push(nextId);
        return ids;
    }
    
    
    this.evaluate = function(expression, vectors) {
        // {'v1': {'refper': "2018-01-01", 'value': 1}, ...}
        expression = expression.replace(/ /g, '')
    
        let infix = splitSymbols(expression);
        let post = postfix(infix);
            
        let stack = [];
        
        for (let s = 0; s < post.length; s++) {
            let symbol = post[s];
            
            if (typeof symbol === 'string' && symbol.startsWith('v')) {
                stack.push(new ExpressionNode(vectors[symbol]));			
            }
            else if (!isNaN(symbol)) {
                stack.push(new ExpressionNode(symbol));
            }
            else {
                let s1 = stack.pop();
                let s2 = stack.pop();
                
                let node = new ExpressionNode(operators[symbol]);
                node.left = s1;
                node.right = s2;
                
                stack.push(node);
            }
        }
        
        return stack.pop().result();
    };
    
    
    ExpressionNode = function(value) {
        this.operation = null;
        this.value = null;  
        this.left = null;
        this.right = null;
        
        if (Array.isArray(value) || !isNaN(value)) {
            this.value = value;
        }
        else {
            this.operation = value;
        }
        
        
        /**
         * Returns a value based on the operation of this node.
        **/
        this.result = function() {
            if (this.isVector() || this.isScalar()) {
                return this.value;
            }  
            else {
                if (this.left == null || this.right == null) {
                    throw new Error('Could not evaluate operator node.'); 
                }
            
                return operate(
                        this.right.result(), this.left.result(), this.operation);
            } 
        }
        
        this.hasChildren = function() {
            return !(this.left == null && this.right == null);
        }
        
        this.isOperator = function() {
            return this.operation != null;
        }

        this.isVector = function() {
            return this.operation == null && Array.isArray(this.value);
        }
        
        this.isScalar = function() {
            return this.operation == null && !isNaN(this.value);
        }
    };
    
    
    /**
     * Returns a vector based on an operation 
     *
     * operation: Function to apply to vector values. 
    **/
    operate = function(valueA, valueB, operation) {
        if (Array.isArray(valueA) && Array.isArray(valueB)) {
            return vectorOperate(valueA, valueB, operation);
        }
        if (Array.isArray(valueA) && !isNaN(valueB)) {
            return vectorScalarOperate(valueA, valueB, operation);
        }	
        if (!isNaN(valueA) && Array.isArray(valueB)) {
            return vectorScalarOperate(valueB, valueA, operation);
        }
        if (!isNaN(valueA) && !isNaN(valueB)) {
            return operation(valueA, valueB);
        }
        
        throw new Error("Unsupported types for operation.");
    };


    vectorScalarOperate = function(vector, scalar, operation) {
        let result = [];
        
        for (let p = 0; p < vector.length; p++) {
            let newPoint = {
                'refper': vector[p].refper,
                'value': operation(vector[p].value, scalar)
            };
            
            // Merge keys added by the user.
            for (key in vector[p]) {
                if (key != 'refper' && key != 'value') {
                    newPoint[key] = vector[p][key];
                }
            }
            
            result.push(newPoint);
        }
        
        return result;
    };


    vectorOperate = function(vectorA, vectorB, operation) {
        // Check if vector lengths match.
        if (vectorA.length != vectorB.length) {
            throw new Error("Vector lengths do not match.");
        }
    
        let result = [];
        
        for (let p = 0; p < vectorA.length; p++) {
            let refperA = vectorA[p].refper;
            let refperB = vectorB[p].refper;
            if (refperA instanceof Date) refperA = refperA.getTime();
            if (refperB instanceof Date) refperB = refperB.getTime();
            
            // Reference period of both vectors points must match.
            if (refperA != refperB) {
                throw new Error("Vectors are not interopable.");
            }
            
            let newPoint =  {
                'refper': vectorA[p].refper, 
                'value': operation(vectorA[p].value, vectorB[p].value)
            };
            
            // Merge keys added by the user.
            for (key in vectorA[p]) {
                if (key != 'refper' && key != 'value') {
                    newPoint[key] = vectorA[p][key];
                }
            }
            
            result.push(newPoint);
        }
        
        return result;
    };


    postfix = function(symbols) {
        let stack = ['('];
        let post = [];
        symbols.push(')');
        
        for (let s = 0; s < symbols.length; s++) {
            let symbol = symbols[s];
            
            if (!isNaN(symbol)) {
                post.push(symbol);
            }
            else if (symbol.startsWith('v')) {
                post.push(symbol);
            }
            else if (symbol == '(') {		
                stack.push('(');
            }	
            else if (symbol == ')') {
                while (stack[stack.length - 1] != '(') {
                    post.push(stack.pop());
                }
                stack.pop();
            } 	
            else {
                while(priority(symbol) <= priority(stack[stack.length - 1])) {
                    post.push(stack.pop());
                }
            
                stack.push(symbol);
            }
        }
        
        return post;
    };


    priority = function(symbol) {
        if (symbol in operatorPriorities) {
            return operatorPriorities[symbol];
        }
        
        return 0;
    };


    splitSymbols = function(vexp) {
        let split = [];
        
        for (let pos = 0; pos < vexp.length; pos++) {
            let next = null;
            
            if (vexp[pos] == 'v' || vexp[pos] == 'V') {
                next = readVector(vexp, pos);
            }
            else if (!isNaN(vexp[pos]) || 
                    (vexp[pos] == '-' && isNaN(vexp[pos - 1]) && !isNaN(vexp[pos + 1]))) {
                next = readScalar(vexp, pos);
            }
            else if (vexp[pos] in operators) {
                next = readOperator(vexp, pos);
            }
            else if (vexp[pos] == '(' || vexp[pos] == ')') {
                next = readBracket(vexp, pos);
            }
            else {
                throw new Error("Unrecognized symbol at position " + pos + ".");
            }
            
            split.push(next.symbol);
            pos = next.pos;
        }

        return split;
    };


    readVector = function(vexp, pos) {
        let symbol = "v";
        pos++;
        
        while(!isNaN(vexp[pos]) && pos < vexp.length) {
            symbol += vexp[pos];
            pos++;		
        }

        return {'symbol': symbol, 'pos': pos - 1};
    };


    readOperator = function(vexp, pos) {
        return {'symbol': vexp[pos], 'pos': pos};
    };


    readScalar = function(vexp, pos) {
        let symbol = "";
        let start = pos;
        
        while ((!isNaN(vexp[pos]) || vexp[pos] == '.' 
                || (vexp[pos] == '-' && pos == start)) 
                && pos < vexp.length) {
            symbol += vexp[pos];	
            pos++;
        }
        
        return {'symbol': Number(symbol), 'pos': pos - 1};
    };


    readBracket = function(vexp, pos) {
        return {'symbol': vexp[pos], 'pos': pos};
    };

    
    validateBrackets = function(vexp) {
        // TODO: Expand on this to also return position of incorrect bracket.
        let stack = [];
        
        for (let c = 0; c < vexp.length; c++) {
            if (vexp[c] == '(') {
                stack.push(1);
            }
            if (vexp[c] == ')') {
                if (stack.length == 0) return false;
                stack.pop();
            }
        }

        return stack.length == 0;
    };
}



let vectors = {
    'v1': [
        {'refper': "2018-01-01", 'value': 1},
        {'refper': "2018-02-01", 'value': 2}
    ],
    'v2': [
        {'refper': "2018-01-01", 'value': 3},
        {'refper': "2018-02-01", 'value': 4}
    ],
    'v3': [
        {'refper': "2018-01-01", 'value': 2},
        {'refper': "2018-02-01", 'value': 2}
    ],
    'v4': [
        {'refper': "2018-01-01", 'value': 1},
        {'refper': "2018-02-01", 'value': 2},
        {'refper': "2018-03-01", 'value': 3}
    ],
    'v5': [
        {'refper': "2018-01-01", 'value': 4},
        {'refper': "2018-02-01", 'value': 5},
        {'refper': "2018-03-01", 'value': 6}
    ],
    'v6': [
        {'refper': "2018-01-01", 'value': 7},
        {'refper': "2018-02-01", 'value': 8},
        {'refper': "2018-03-01", 'value': 9}
    ]
};

let testcases = [
    {
        'expression': "(v1 + v2) * (2*v3)",
        'expected': [
            {'refper': "2018-01-01", 'value': 16},
            {'refper': "2018-02-01", 'value': 24}
        ]
    },
    {
        'expression': "(v1 - v2) * (2*v3)",
        'expected': [
            {'refper': "2018-01-01", 'value': -8},
            {'refper': "2018-02-01", 'value': -8}
        ]
    },
    {
        'expression': "v6 - v5 - v4",
        'expected': [
            {'refper': "2018-01-01", 'value': 2},
            {'refper': "2018-02-01", 'value': 1},
            {'refper': "2018-03-01", 'value': 0}
        ]
    }
];

let runTest = function(testcase) {
    let parser = new VectorExpressionEvaluator();
    let result = parser.evaluate(testcase.expression, vectors);
    
    console.log(testcase.expression);
    console.log("Result:");
    console.log(result);
    console.log("Expected:");
    console.log(testcase.expected);
    
    if (parser.equals(result, testcase.expected)) {
        console.log("PASS");
        return true;
    }
    else {
        console.log("FAIL");
        return false;
    }
}

let passes = 0;
for (let t = 0; t < testcases.length; t++) {
    if (runTest(testcases[t])) {
        passes++;
        console.log("===================");
    } 
}
console.log(passes + " / " + testcases.length + " tests passed.");